modules/llrt_crypto/src/subtle/generate_key.rs

```diff
@@ -0,0 +1,294 @@
+            "private".to_string(),
+            extractable,
+            algorithm.clone(),
+            private_usages,
+            &bytes, // for test
+        )?;
+        let public_key = CryptoKey::new(
+            ctx.clone(),
+            "public".to_string(),
+            true,
+            algorithm,
+            public_or_secret_usages,
+            &bytes, // for test
+        )?;
+        CryptoKeyPair::new(ctx.clone(), private_key, public_key).into_js(&ctx)
+    }
+}
+
+fn extract_generate_key_algorithm(
+    ctx: &Ctx<'_>,
+    algorithm: &Value,
+) -> Result<(String, KeyGenAlgorithm)> {
+    let name = algorithm
+        .get_optional::<_, String>("name")?
+        .ok_or_else(|| Exception::throw_type(ctx, "algorithm 'name' property required"))?;
+
+    match name.as_str() {
+        "RSASSA-PKCS1-v1_5" | "RSA-PSS" | "RSA-OAEP" => {
+            let modulus_length = algorithm.get_optional("modulusLength")?.ok_or_else(|| {
+                Exception::throw_type(ctx, "algorithm 'modulusLength' property required")
+            })?;
+
+            let public_exponent = algorithm.get_optional("publicExponent")?.ok_or_else(|| {
+                Exception::throw_type(ctx, "algorithm 'publicExponent' property required")
+            })?;
+
+            Ok((name, KeyGenAlgorithm::Rsa { modulus_length, public_exponent }))
+        },
+        "ECDSA" | "ECDH" => {
+            let named_curve = algorithm
+                .get_optional::<_, String>("namedCurve")?
+                .ok_or_else(|| Exception::throw_type(ctx, "algorithm 'namedCurve' property required"))?;
+
+            let curve = CryptoNamedCurve::try_from(named_curve.as_str()).or_throw(ctx)?;
+
+            Ok((name, KeyGenAlgorithm::Ec { curve }))
+        },
+        "HMAC" => {
+            let hash = extract_sha_hash(ctx, algorithm)?;
+
+            let length = algorithm.get_optional::<_, u32>("length")?;
+
+            Ok((name, KeyGenAlgorithm::Hmac { hash, length }))
+        },
+        "AES-CTR" | "AES-CBC" | "AES-GCM" | "AES-KW" => {
+            let length = algorithm
+                .get_optional("length")?
+                .ok_or_else(|| Exception::throw_type(ctx, "algorithm 'length' property required"))?;
+
+            if length != 128 && length != 192 && length != 256 {
+                return Err(Exception::throw_type(
+                    ctx,
+                    "Algorithm 'length' must be one of: 128, 192, or 256",
+                ));
+            }
+
+            Ok((name, KeyGenAlgorithm::Aes { length }))
+        },
+        _ => Err(Exception::throw_message(
+            ctx,
+            "Algorithm 'name' must be RsaHashedKeyGenParams | EcKeyGenParams | HmacKeyGenParams | AesKeyGenParams",
+        )),
+    }
+}
+
+fn classify_and_check_usages<'js>(
+    ctx: &Ctx<'js>,
+    name: &str,
+    key_usages: &Array<'js>,
+) -> Result<(Array<'js>, Array<'js>)> {
+    fn get_usages(arr: &[(&str, &[&str])], algorithm: &str) -> Option<HashSet<String>> {
```

Comment:
We can simplify here by not returning hashset but instead working with arrays. Remember that HashSet is great for more than 20-50+ items. Otherwise it's slower. We also can return the referenced array, no need to create a copy.
```rust
fn find_usages<'a>(table: &'a [(&str, &[&str])], algorithm: &str) -> Option<&'a [&'a str]> {
    table
        .iter()
        .find(|(name, _)| *name == algorithm)
        .map(|(_, usages)| *usages)
}
```

Then we can instead use a single for loop, something like:
```rust
let supported_usages = find_usages(SUPPORTED_USAGES_ARRAY)
let mandatory_usages = find_usages(MANDATORY_USAGES_ARRAY)
let mut private_usages: Vec<String> = Vec::with_capacity(key_usages_set.len());
let mut public_usages: Vec<String> = Vec::with_capacity(key_usages_set.len());

for usage in key_usages_set {
    if !supported_usages.contains(&usage.as_str()) {
        return Err(...);
    }
    if mandatory_usages.contains(&usage.as_str()) {
        private_usages.push(usage.into());
    }else{
        public_usages.push(usage.into());
    }
}
if !supported_usages.is_empty() && private_usages.is_empty(){
  return Err(...) //we have no mandatory usages
}
```